adcm_client/packer/types.py

```diff
@@ -139,71 +161,93 @@ def _get_prepared_image(pkgs, image: Image, client: DockerClient) -> "Image":
-    with open(os.path.join(source_path, kwargs['requirements']), 'r') as file:
-        pkgs = yaml.safe_load(file)
-        if pkgs.get('python_mod'):
-            client = from_env()
-            # choose image where to install python pkgs
-            # by default adcm:latest but i think this may be bad practice
-            # better to use adcm_min_version of bundle
-            image_name = "arenadata/adcm:latest" if not kwargs.get('image') else kwargs['image']
-            image = client.images.pull(image_name)
-            # clean up flag
-            rm_prepared_image = True
-
-            # prepared image is an image with intalled python packages
-            # that need to be packed in bundle
-            # for debug purposes there may be an prepared_image variable received from spec.yaml
-            if kwargs.get('prepared_image'):
-                try:
-                    prepared_image = client.images.get(kwargs['prepared_image'])
-                    rm_prepared_image = False
-                except ImageNotFound:
-                    prepared_image = _get_prepared_image(pkgs, image, client)
-            else:
-                prepared_image = _get_prepared_image(pkgs, image, client)
-
-            # list of all highlevel dirs of python pkgs that must be packed
-            dirs = _get_top_dirs(image, prepared_image, client)
-
-            # volume that contains workspace
-            volumes = {
-                workspace: {'bind': workspace, 'mode': 'rw'}
-            }
-
-            if kwargs.get('target_dir'):
-                path = os.path.join(source_path, kwargs['target_dir'], 'pmod')
-            else:
-                path = os.path.join(source_path, 'pmod')
-            _copy_pkgs_files(path, dirs, prepared_image, volumes, client)
-
-            if rm_prepared_image:
-                client.images.remove(prepared_image.id)
-
-        else:
-            raise NoModulesToInstall('Can`t get python modules list to be inatalled')
+def _file_hash(data: str) -> blake2b:
+    return blake2b(bytes(data, "utf-8"), digest_size=4)
+
+
+def _prepared_image_name(base_image: str, hash_obj: blake2b) -> str:
+    hash_obj.update(bytes(base_image, "utf-8"))
+    return f"adcm_sdk_pack/python_mod_req_adcm:{hash_obj.hexdigest()}"
+
+
+def python_mod_req(
+    source_path, workspace, *, flags: PackFlags = None, target_dir: str = "", **kwargs
+):
+    with open(os.path.join(source_path, kwargs["requirements"]), "r") as file:
+        data = file.read()
+        req_file_hash = _file_hash(data)
+        pkgs = yaml.safe_load(data)
+
+    if "python_mod" not in pkgs:
+        raise NoModulesToInstall("Can`t get python modules list to be inatalled")
+
+    client = from_env()
+    # choose image where to install python pkgs
+    # by default adcm:latest but i think this may be bad practice
+    # better to use adcm_min_version of bundle
+    image_name = kwargs.get("image", "arenadata/adcm:latest")
```

Comment:
```suggestion
    image_name = kwargs.get("image", "hub.arenadata.io/adcm/adcm:latest")
```

Reply:
@dgusakov I removed this variable. Using a default image instead of defined one will lead to a unpredictable result in long terms.